added SimpleTypeNameSerializer for better generic typename handling

heinzbeinz opened the pull request:
Hi,

since I had a problem with queue names of generic types getting too long, I wrote a custom type name serializer. Maybe you find it useful, too.

Example typename:
Shared.Command.ControlServer.Location.ProxyCommandPayload`1[[Shared.Command.Generic.Reflection.ReflectCommandPayload, Shared]], Shared

Pliner commented:
Hi @heinzbeinz, 

Thank you for contribution.

From my point of view there are two things to discuss:
1. Parsing... To parse connection string in EasyNetQ we use [Sprache](https://github.com/sprache/Sprache). Could you try it instead of manual parsing?
2. Making this serializer by default. Maybe it is possible to make it by default and have backward compability with previous one?

micdenny commented:
> 2. Making this serializer by default. Maybe it is possible to make it by default and have backward compability with previous one?

I think we must maintain the previous one as default, or at least as an automatic fallback (dunno how to now)

heinzbeinz commented:
Hi @Pliner and @micdenny,

thanks for your reply. Unfortunately I didn't manage to create a Sprache-based parser as I failed to create the required recursion. I created a EBNF grammar for it:

    TYPECHAR = LETTER | DIGIT | "." | "`" .
    TYPENAME = TYPECHAR  { TYPECHAR } .
    GENTYPE  = "[" TYPE "]" .
    TYPE     = TYPENAME [ "[" GENTYPE { "," GENTYPE } "]" ] "," TYPENAME .

Making this parser the default would break backward compatibility for generic types. If backward compatiblity is a necessity, then we could prefix the typestring to select the appropriate parser (no prefix -> default, with prefix -> this parser).

kjellski commented:
*bump* 
This is a really nice workarround, and Sprache is not capable of generating a parser for this, how would one proceed now?

micdenny commented:
@heinzbeinz can you add unit tests also for your new `SimpleTypeNameSerializer`?

heinzbeinz commented:
Hi @micdenny,

I added the unit tests (created a new project EasyNetQ.Tests.NET45)

Pliner commented:
@heinzbeinz @micdenny @kjellski 
Imho, it's possible to create sprache based parser. Here is (my attempt)[https://gist.github.com/Pliner/281558af47d29bfcaab1] to do it.

Could you review it?

heinzbeinz commented:
@Pliner 

thanks for your parser, but it seems to have a problem with nested types like Dictionary<string, List<int>>. The following string cannot be parsed: 

System.Collections.Generic.Dictionary`2[[System.String, mscorlib], [System.Collections.Generic.List`1[[System.Int32, mscorlib]], mscorlib]], mscorlib

Pliner commented:
Thanks @heinzbeinz :wink:

It was just a sketch of solution, because I had'n time...
Here is an updated version of parser https://gist.github.com/Pliner/281558af47d29bfcaab1

heinzbeinz commented:
Hi @Pliner,

thanks for your update! Your parser works like a charme. I implemented it like https://gist.github.com/heinzbeinz/32c5ba1e9c8121632cf7

But now I have a problem integrating it into EasyNetQ. In EasyNetQ there is no nuget package for Sprache. It is integrated in the core via source code, but this version seems outdated (e.g. no Optional() is available). I tried to reference the nuget package (and replaced CaseInsensitiveString with IgnoreCase), but afterwards some Unittests fail... maybe the parser semtantics changed.

Pliner commented:
@heinzbeinz  I could try to replace Sprache code with nuget and ilmerge it.

@micdenny What do you think?

heinzbeinz commented:
@Pliner that would be great! Thanks.

micdenny commented:
@Pliner I like to have this `SimpleTypeNameSerializer` that can be chosen instead of the actual one `TypeNameSerializer` and probably would also fixed problems like the one described here #562 

@Pliner would you like to conclude this PR using `Sprache` as suggested by @heinzbeinz? or we close as it now and then open an issue to give it a try with `Sprache`?

Pliner commented:
Will be fixed by #786
